### PR TITLE
net: lwm2m: Add Event Log object

### PR DESCRIPTION
**Pull request by RomainPelletant**
Add lwm2m standard object for event log (id: 20)

OMA ressource object can be found here : https://devtoolkit.openmobilealliance.org/OEditor/LWMOView?url=https%3A%2F%2Fraw.githubusercontent.com%2FOpenMobileAlliance%2Flwm2m-registry%2Fprod%2F20.xml

Signed-off-by: Romain Pelletant <romainp@kickmaker.net>

### CODE REVIEW

File: include/zephyr/net/lwm2m.h [hunk 1]
```diff
@@ -559,6 +560,20 @@ int lwm2m_swmgmt_install_completed(uint16_t obj_inst_id, int error_code);
...
+
+/**
+ * @brief Set callback to read log data
+ *
+ * The callback will be executed when the LWM2M read operation gets called
+ * on the corresponding object.
+ *
+ * @param[in] cb A callback function for handling the read event.
+ */
+void lwm2m_event_log_set_read_log_data_cb(lwm2m_engine_get_data_cb_t cb);
```
Comment (rlubos) [inline include/zephyr/net/lwm2m.h:573]: Do we really need a custom function for this, couldn't the application simply use `lwm2m_engine_register_read_callback()` to register a read callback?

File: subsys/net/lib/lwm2m/Kconfig [hunk 2]
```diff
@@ -500,6 +500,11 @@ config LWM2M_LOCATION_OBJ_SUPPORT
 config LWM2M_GATEWAY_OBJ_SUPPORT
 	bool "LwM2M Gateway Object (25) [EXPERIMENTAL]"
 
+config LWM2M_EVENT_LOG_OBJ_SUPPORT
+	bool "LwM2M Event Log Object (20) [EXPERIMENTAL]"
+	help
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:505]: Experimental Kconfigs should have `select EXPERIMENTAL`  entry.

File: subsys/net/lib/lwm2m/lwm2m_obj_event_log.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+struct lwm2m_event_log_obj {
+	uint8_t log_class;
+	uint8_t log_status;
+	uint8_t log_data_format;
+};
+
+static struct lwm2m_engine_obj lwm2m_event_log;
+static struct lwm2m_engine_obj_field fields[] = {
+	OBJ_FIELD_DATA(LWM2M_EVENT_LOG_CLASS_ID, R_OPT, U8),
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_event_log.c:47]: The object spec says this should be a RW resource.

File: subsys/net/lib/lwm2m/lwm2m_obj_event_log.c [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+static int lwm2m_event_log_init(const struct device *dev)
+{
+	struct lwm2m_engine_obj_inst *obj_inst = NULL;
+	int ret = 0;
+
+	/* initialize the Event Log field data */
+	lwm2m_event_log.obj_id = LWM2M_OBJECT_EVENT_LOG_ID;
+	lwm2m_event_log.version_major = EVENT_LOG_VERSION_MAJOR;
+	lwm2m_event_log.version_minor = EVENT_LOG_VERSION_MINOR;
+	lwm2m_event_log.is_core = true;
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_event_log.c:109]: This is not a "core" object - i. e. not part of the LwM2M Technical Specification.

**Review by rlubos — COMMENTED** [review]
Thank you for the contributions, a few comments from my side.

Please note, that you need to add some description in the commit message in order to pass the compliance.

File: subsys/net/lib/lwm2m/Kconfig (hunk 2 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/Kconfig:505]: Remove experimental, no needed in that case

File: subsys/net/lib/lwm2m/lwm2m_obj_event_log.c (hunk 3 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_obj_event_log.c]: You are totally right: add write attribute

File: subsys/net/lib/lwm2m/lwm2m_obj_event_log.c (hunk 4 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_obj_event_log.c]: Set to false

**Review by rlubos — COMMENTED** [review]
Just please remove the `[EXPERIMENTAL]` from the commit title, and from my side we're good.

**Comment by RomainPelletant** [comment]
> Just please remove the `[EXPERIMENTAL]` from the commit title, and from my side we're good.

Done, thank you for the review

**Review by rlubos — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**